### PR TITLE
SpinWidget: enable OK button only for changed value

### PR DESCRIPTION
**Pull request by hius07**
`Apply` button is enabled only after changing the value in the NumberPicker or with the Default button.
While changing, if the value in the NumberPicker becomes equal to the initial value, `Apply` button goes disabled again.

Inspired by https://github.com/koreader/koreader/pull/8490#discussion_r759157519.

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/8495)

### CODE REVIEW

File: frontend/ui/widget/spinwidget.lua [hunk 1]
```diff
@@ -52,6 +52,7 @@ local SpinWidget = InputContainer:new{
 }
 
 function SpinWidget:init()
+    self._value = self.value -- used to enable ok_button, self.value may be changed in extra callback
```
Comment (poire-z) [inline frontend/ui/widget/spinwidget.lua:55]: You could name it with an explicite name `self.currently_applied_value` or `self.original_value`.
It would make the stuff below reads as natural language :)

**Review by poire-z — APPROVED** [review]
Small and neat.

**Comment by poire-z** [comment]
If others agree with the idea.
Does it just work:tm: with the bottom config spin widgets, ie Font size ?

**Comment by hius07** [comment]
Yeah, took a lot of time for testing configdialog. There is a bug there, which made me nervous of my code, but okay then.

https://github.com/koreader/koreader/blob/c7a3c17ce06668137e6bcb231315937a401e5abc/frontend/ui/widget/configdialog.lua#L1317

If press `Set as default`, then do not close the SpinWidget, press it again, the new default value is not saved.
Got in the defaul.settings.lua
```
["copt_copt_copt_copt_font_size"] = 22,
["copt_copt_copt_font_size"] = 23,
["copt_copt_font_size"] = 22.5,

```
The prefix is added more and more. To be fixed separately.

**Comment by zwim** [comment]
:+1 from me.

**Comment by poire-z** [comment]
Now that your other DoubleSpinWidget is merged, you could tweak it similarly - in this PR or another one?

**Comment by hius07** [comment]
In this one, if you don't mind.

**Comment by hius07** [comment]
Done for DSW.

File: frontend/ui/widget/doublespinwidget.lua [hunk 2]
```diff
@@ -221,11 +229,16 @@ function DoubleSpinWidget:update()
         },
         {
             text = self.ok_text,
+            enabled = self.left_value ~= left_widget:getValue() or self.right_value ~= right_widget:getValue(),
             callback = function()
                 if self.callback then
-                    self.callback(left_widget:getValue(), right_widget:getValue())
+                    self.left_value = left_widget:getValue()
+                    self.right_value = right_widget:getValue()
```
Comment (poire-z) [inline frontend/ui/widget/doublespinwidget.lua:236]: No need to save the values / update self.left/right_value when there is no `self.callback` ?

Comment (hius07) [reply frontend/ui/widget/doublespinwidget.lua:236]: > when there is no self.callback

I just wanted to ask, what is the reason of this check? How can be a spinwidget without callback on ok_button?

Comment (poire-z) [reply frontend/ui/widget/doublespinwidget.lua:236]: Dunno, I don't have all the answers :)
May be it is just a safety check.
May be it was based from some other widget that has an optional callback.
It could even be used as a read only display stuff (like Book status) (that I know we don't have that use any spin widget).

(Usually, I don't question what's there, I just follow the existing logic, try to fit what I add in what's already there :)  I rarely feel the need to clean up, rationalize - with the risk involved of breaking things :)

Your current code would mean that self.left_value and right_value has no use when there is no self.callback - which is fine if that's really the case.

Comment (hius07) [reply frontend/ui/widget/doublespinwidget.lua]: Let's update self values on pressing ok_button, with or without self.callback.

**Review by poire-z — APPROVED** [review]

File: frontend/ui/widget/doublespinwidget.lua [hunk 3]
```diff
@@ -102,20 +102,28 @@ function DoubleSpinWidget:update()
...
         precision = self.right_precision,
         wrap = self.right_wrap,
     }
+    left_widget.picker_updated_callback = function(value)
+        self:update(value, right_widget:getValue())
+    end
+    right_widget.picker_updated_callback = function(value)
+        self:update(left_widget:getValue(), value)
+    end
+
```
Comment (poire-z) [inline frontend/ui/widget/doublespinwidget.lua:119]:
```
Checking frontend/ui/widget/doublespinwidget.lua  1 warning
    frontend/ui/widget/doublespinwidget.lua:119:1: line contains only whitespace
Total: 1 warning / 0 errors in 474 files
```
To be fixed on your next PR (no need to make a dedicated PR, it is minor, and you'll make a real PR soon enough :)
